### ISSUE TITLE
How to train the generator ?

### ISSUE THREAD

**Issue by LazerLikeFocus** [black0017/3D-GAN-pytorch#3]
https://github.com/black0017/3D-GAN-pytorch/blob/e3f640dbb8335cde239334b3b1ad143acd784c56/train_gans.py#L53

has only the discriminator model inside it. 
how does the generator gets updated ?

I know the fake_data are created by the generator but how to update the generator model when its not even in the above funct ?

in keras we used to create a separate gan function combining gen + dis models sequentially , 
and make the dis non-trainable and later update the gen using the gan's error

def define_gan(g_model, d_model):

	d_model.trainable = False                  # make weights in the discriminator not trainable

	model = Sequential()                         # connect them

	model.add(g_model)                         # add generator
                                   
	model.add(d_model)                        # add the discriminator 

	opt = Adam(lr=0.0002, beta_1=0.5) 
	model.compile(loss='binary_crossentropy', optimizer=opt)
	return model

**Comment by black0017** [comment]
hi, thanks for the issue!

please take a look here https://github.com/black0017/3D-GAN-pytorch/issues/1

and let me know if you managed to get it to work.

in short here is the logic:
```python
# 1. Train Discriminator
latent_space_data = noise(args.batch_size)
if args.cuda:
    latent_space_data = latent_space_data.cuda()
    real_data = real_data.cuda()

    fake_data = generator(latent_space_data).detach()

    d_error, d_pred_real, d_pred_fake = train_discriminator(
        discriminator, d_optimizer, real_data, fake_data,
        loss_d, epoch, args.cuda)

# 2. Train Generator
latent_space_data = noise(args.batch_size)
if args.cuda:
    latent_space_data = latent_space_data.cuda()

fake_data = generator(latent_space_data)
g_error = train_generator(discriminator, g_optimizer, fake_data, loss_g, args.cuda) 
```
Maybe i should add another function that capture the whole train logic as in the referenced issue issue?

best,

N.A.

**Comment by LazerLikeFocus** [comment]
`g_optimizer = optim.Adam(generator.parameters(), args.lr_g, betas=args.betas)`

What is model parameters () in PyTorch?
In PyTorch, the learnable parameters (i.e. **weights and biases**) of a torch. nn. Module model are contained in the model's parameters (accessed with model. parameters() ). 

`train_generator(discriminator, g_optimizer, fake_data, loss_g, args.cuda) `
g_optimizer does the trick.

Im from keras background and new to pytorch.
Keras is way different than pytorch.

Anyway thank you for quick reply.

**Comment by black0017** [comment]
Ah ok I see the point now. You answered better to yourself than me :)
Best!
